Core/include/Acts/Vertexing/NumericalTrackLinearizer.ipp

```diff
@@ -0,0 +1,143 @@
+    linearizeTrack(const BoundTrackParameters& params, const Vector4& linPoint,
+                   const Acts::GeometryContext& gctx,
+                   const Acts::MagneticFieldContext& mctx) const {
+  // Make Perigee surface at linPointPos, transverse plane of Perigee
+  // corresponds the global x-y plane
+  Vector3 linPointPos{VectorHelpers::position(linPoint)};
+  std::shared_ptr<PerigeeSurface> perigeeSurface{
+      Surface::makeShared<PerigeeSurface>(linPointPos)};
+
+  // Create propagator options
+  propagator_options_t pOptions(gctx, mctx);
+
+  // Length scale at which we consider to be sufficiently close to the Perigee
+  // surface to skip the propagation.
+  pOptions.targetTolerance = m_cfg.targetTolerance;
+
+  // Get intersection of the track with the Perigee if the particle would
+  // move on a straight line.
+  // This allows us to determine whether we need to propagate the track
+  // forward or backward to arrive at the PCA.
+  auto intersection = perigeeSurface->intersect(gctx, params.position(gctx),
+                                                params.unitDirection(), false);
+
+  // Setting the propagation direction using the intersection length from
+  // above.
+  // We handle zero path length as forward propagation, but we could actually
+  // skip the whole propagation in this case.
+  pOptions.direction =
+      Direction::fromScalarZeroAsPositive(intersection.intersection.pathLength);
+
+  // Propagate to the PCA of linPointPos
+  auto result = m_cfg.propagator->propagate(params, *perigeeSurface, pOptions);
+
+  // Extracting the Perigee representation of the track wrt linPointPos
+  auto endParams = *result->endParameters;
+  BoundVector perigeeParams = endParams.parameters();
+
+  // Covariance and weight matrix at the PCA to "linPoint"
+  BoundSymMatrix parCovarianceAtPCA{endParams.covariance().value()};
+  BoundSymMatrix weightAtPCA{parCovarianceAtPCA.inverse()};
+
+  // Vector containing curvilinear track parameters at the PCA
+  const unsigned int nCurvilinearParams{7};
+  Acts::ActsVector<nCurvilinearParams> curvilinearParamVec;
+
+  // 4D PCA and the momentum of the track at the PCA
+  // These quantities will be used in the computation of the constant term in
+  // the Taylor expansion
+  Vector4 pca;
+  Vector3 momentumAtPCA;
+
+  // Fill "curvilinearParamVec", "pca", and "momentumAtPCA"
+  {
+    Vector3 globalCoords{endParams.position(gctx)};
+    double globalTime{endParams.time()};
+    double phi{perigeeParams(BoundIndices::eBoundPhi)};
+    double theta{perigeeParams(BoundIndices::eBoundTheta)};
+    double qOvP{perigeeParams(BoundIndices::eBoundQOverP)};
+
+    curvilinearParamVec << globalCoords, globalTime, phi, theta, qOvP;
+    pca << globalCoords, globalTime;
+    momentumAtPCA << phi, theta, qOvP;
+  }
+
+  // Setting size of the perturbation delta for calculation of numerical
+  // derivatives (i.e., f'(x) ~ (f(x+delta) - f(x)) / delta)
+  double delta{1e-8};
```

Comment:
I am trying to understand how the Linearizers are working so my question would be dumb.
isn't delta should be `intersection.intersection.pathLength` instead?

Reply:
Or the linearizer is supposed to calculate the derivative at the fixed point rather than calculating the transport jacobian

Reply:
the delta is how much we change the initial parameters and then we propagate to the PCA again which results in different final parameters. afterwards @russofel determines the numerical derivatives by calculating `(final - initial) / delta`

using the transport jacobian should give a similar result I guess?

I think the job of the linearizer is just to provide a set of matrices that can be used in the fitter to linearly explore the parameter space in order to be able to update your prediction with the "measurement"

Reply:
The only thing `intersection.intersection.pathLength` does is that it tells us in which direction to go when we propagate to a point!

Yes, we calculate the derivative at a fixed point (i.e., the PCA to a reference point in our case). 

Concerning the transport Jacobian: Bastian hints in his thesis that the transport Jacobian should correspond to our result if we let the path length between the two points go to 0 (see Fig. 5.7 in https://cds.cern.ch/record/2839800). I am not entirely sure how this works, but I will look into it.

Reply:
OK. I looked into the figure and corresponding text - If it is just a jacobian transformation on a fixed point, the approach in Numerical linearizer seems correct.

Now I don't understand why HelicalTrackLinearizer is calculating two cases (b field and no b field) differently when we assume that there is no step length :p . The way I see is that the jacobian is simple a coordinate transformation from perigee to free parameter without transport jacobian

Reply:
I think the Jacobian we have here is not just a coordinate transformation: It is the derivative of the Perigee parametrization with respect to the free parameters at the PCA. 

If we go to the PCA and slightly wiggle one of the track parameters there, the track will change and we will get a new PCA. The Jacobian now comprises the following information: "How do the parameters at the new PCA change compared to the old one?"

I hope this answers why we must treat the cases of no B-field and constant B-field differently: The new PCA will depend on the magnetic field - or the lack thereof!

Reply:
hm I think the jacobian should account for all of that. is the transport really just identity? maybe there are some terms that result into something else when you take the limit at the very end of the calculation? would be interesting to see

@russofel could you link some material were you calculated the derivatives? did you plug in the helix equations or how did that work?

Reply:
Yeah Transport jacobian converges to the identity when delta (or step length) is closed to zero.

Reply:
> It might not have nothing to do with this PR, but there is also a way to validate `HelicalTrackLinearizer` with the analytical solution of a helix model.
> 
> See https://github.com/acts-project/detray/blob/51571670ea7b8e574db560c9dadf49401352583d/core/include/detray/intersection/detail/trajectories.hpp#L268
> 
> which is based on the helix model (Eq. 4.28 - 4.34) of https://doi.org/10.1007/978-3-030-65771-0

The HelicalTrackLinearizer is purely based on analytical calculations, see Sec. 5.3.3 and 5.3.4 of https://cds.cern.ch/record/1243771?ln=en. 

I am not quite sure how these calculations relate to the resource you have linked in your comment though... I will need to look into this in detail!